### DISCUSSION TITLE
Get Request/Response from DevToolClient

### DISCUSSION THREAD

**Discussion by bbday** [cefsharp/CefSharp#3652] [General]
I'm try to get request / response from devtoolclient

using (var devToolsClient = _browser.GetDevToolsClient())
devToolsClient.Network [...]

But i didnt find anything to collect http request/response from browser, i checked also here http://cefsharp.github.io/api/91.1.x/html/N_CefSharp_DevTools_Network.htm no clue

Can you provide a small example? thanks.

**Comment by totti240282** [comment]
also i am trying to filter traffic as if i were on browser but i can't.

**Reply by amaitland** [reply]
Using ResourceHandler/ResourceRequestHandler is the recommend approach  search https://github.com/cefsharp/CefSharp/wiki/General-Usage for general usage examples. 

http://cefsharp.github.io/api/91.1.x/html/T_CefSharp_Handler_ResourceRequestHandler.htm

**Reply by totti240282** [reply]
yes I use that, but I wanted to know how to do with the tools.

**Comment by amaitland** [comment]
You can browse/search the DevTools API using https://chromedevtools.github.io/devtools-protocol/tot/Network/

I believe request/response are event based. There's currently no strongly typed wrapper for the events. You can parse the raw JSON yourself. 

http://cefsharp.github.io/api/91.1.x/html/M_CefSharp_Callback_IDevToolsMessageObserver_OnDevToolsEvent.htm
Look at the source for usage example.

**Reply by amaitland** [reply]
Strongly typed events will be added in #3666

**Comment by campersau** [comment]
Here is a small example for logging requests / responses via the DevTools events based on the work in https://github.com/cefsharp/CefSharp/pull/3669.

```c#
using (var browser = new ChromiumWebBrowser("about:blank", automaticallyCreateBrowser: false))
{
    await browser.CreateBrowserAsync();

    using (var devToolsClient = browser.GetDevToolsClient())
    {
        await devToolsClient.Network.EnableAsync(); // enable network events

        devToolsClient.Network.RequestWillBeSent += (sender, args) =>
        {
            Console.WriteLine("--- RequestWillBeSent ---");
            Console.WriteLine("Request: " + args.RequestId);
            Console.WriteLine(args.Request.Method + " " + args.Request.Url + args.Request.UrlFragment);
            foreach (var header in args.Request.Headers)
            {
                Console.WriteLine("Request Header: " + header.Key + " = " + header.Value);
            }
            Console.WriteLine("--------------------------------");
        };
        devToolsClient.Network.RequestWillBeSentExtraInfo += (sender, args) =>
        {
            Console.WriteLine("--- RequestWillBeSentExtraInfo ---");
            Console.WriteLine("Request: " + args.RequestId);
            foreach (var header in args.Headers)
            {
                Console.WriteLine("Request Header: " + header.Key + " = " + header.Value);
            }
            Console.WriteLine("--------------------------------");
        };
        devToolsClient.Network.ResponseReceived += (sender, args) =>
        {
            Console.WriteLine("--- ResponseReceived ---");
            Console.WriteLine("Request: " + args.RequestId);
            if (args.Response.RequestHeaders != null)
            {
                foreach (var header in args.Response.RequestHeaders)
                {
                    Console.WriteLine("Request Header: " + header.Key + " = " + header.Value);
                }
            }
            Console.WriteLine("Response: " + args.Response.Url + " " + args.Response.Status + " " + args.Response.StatusText);
            Console.WriteLine("Connection: " + args.Response.RemoteIPAddress + ":" + args.Response.RemotePort);
            foreach (var header in args.Response.Headers)
            {
                Console.WriteLine("Response Header: " + header.Key + " = " + header.Value);
            }
            Console.WriteLine("--------------------------------");
        };
        devToolsClient.Network.ResponseReceivedExtraInfo += (sender, args) =>
        {
            Console.WriteLine("--- ResponseReceivedExtraInfo ---");
            Console.WriteLine("Request: " + args.RequestId);
            foreach (var header in args.Headers)
            {
                Console.WriteLine("Response Header: " + header.Key + " = " + header.Value);
            }
            Console.WriteLine("--------------------------------");
        };

        await browser.LoadUrlAsync("www.google.com");
    }
}
```
This just logs a few fields, there is a lot more data especially in the responses like timings.

You can use `devToolsClient.Network.GetResponseBodyAsync()` to get the response body for a `RequestId` (can throw if there is none, e.g. for redirects).

---

I haven't figured out how request interception works with the DevTools yet. A lot of the types / events are currently experimental or already deprecated.

**Comment by hhxdestiny** [comment]
RequestWillBeSent undefine

**Reply by amaitland** [reply]
If you look at the linked pull request above you'll see it's attached to the 92 milestone. Basically strongly typed events will be part of the 92 release. 

You can try a CI build available on the myget feed. Important to remember these are development builds and generally not ready for mainstream use.

https://www.myget.org/gallery/cefsharp
